### ISSUE TITLE
DirichletBC ignored when assemblying matrix on two copies of the same FunctionSpace

### ISSUE THREAD

**Issue by francesco-ballarin** [FEniCS/dolfinx#540]
Dear all,
I slightly changed the Taylor-Hood test in python/test/unit/fem/test_assembler.py at
https://github.com/francesco-ballarin/dolfinx/commit/7fd42a5d15d17daca619fe679c7a4797fa3a6395
to show an unexpected behavior (at least to me) when assemblying matrices.

It seems that when a form defined with test/trial on two different function spaces is assembled, DirichletBC are silently ignored (I verified this exporting the matrix content before and after my commit).
While I get that it may not be relevant to apply DirichletBCs if the two function spaces are indeed different, I find it odd that DirichletBC application is ignored even if the two function spaces are actually a copy of the same one.

I am opening this to inquire whether this stricter check is an intended behavior, or if this not the desired one.

**Comment by garth-wells** [comment]
The issue here is that the function spaces are different in the example code. It's important for handling block systems properly that we know if the test/trial functions come from the same space as this allow us to identify diagonal or off-diagonal blocks. In the case posted they could well off-diagonal blocks. Diagonal and off-diagonal blocks need different treatment for Dirichlet bcs.

**Comment by michalhabera** [comment]
I was talking about this with @blechta . It seems, that current interface is wrong in accepting just a list of DirichletBCs. Wouldn't it be better to accept a list of lists of DirichletBCs instead?

`assemble_matrix(..., [[bcs for row block 0], [bcs for row block 1], ...])`

**Comment by francesco-ballarin** [comment]
> The issue here is that the function spaces are different in the example code. It's important for handling block systems properly that we know if the test/trial functions come from the same space as this allows us to identify diagonal or off-diagonal blocks. In the case posted they could well off-diagonal blocks. Diagonal and off-diagonal blocks need different treatment for Dirichlet bcs.

OK, I understand. Feel free to close if this is indeed the intended behavior.

**Comment by garth-wells** [comment]
@michalhabera Possible, but it doesn't really deal with the reported issue. What about for columns? It's the relationship between the function spaces that makes the whole thing tick. 

We could give more structure to how bcs are passed. At the moment the users can throw the whole lot in and the assembler figures out what bcs to apply.

**Comment by blechta** [comment]
The current behavior seems to be suboptimal. BCs are matched to blocks by function space comparison, if I understand it correctly. So it is not very clear what happens when row space or column space contains two or more copies of the same space. This is a legitimate use case in my opinion, as it would be wasting memory creating the same (before BCs) space multiple times.

I would strongly advocate using much more explicit interface. It seems to me that this matches the philosophy of dolfinx.

It is true that the approach @michalhabera mentioned did not think of off-diagonal block assembly. It needs more thought.

**Comment by garth-wells** [comment]
The data sharing is an issue that can be dealt with easily - a new space can share the data and so avoid data duplication. This is in place already at a low level. Whether or not a high level interface is in place would need to checked.

The key point is that it all slots into place if every block row has an associated unique function space object, and every block column has an associated function space. The spaces can easily share data. We then know if a block if off-diagonal or diagonal, and which Dirichlet bcs get applied to which rows and colums.

**Comment by francesco-ballarin** [comment]
I agree with the comment of @blechta .

**Comment by garth-wells** [comment]
It's not clear to me what points there are to agree with. We need to be very clear on the current state and the needs to frame a discussion for changes. As I see it:

1. The sharing of data for `FunctionSpaces` is straightforward and already possible. It may be that a higher level interface would be helpful.
2. There seems to be a suggestion to break the link between a `DirichletBC` and the `FunctionSpace` that it constrains. This would a major change from the past design (and one which hasn't caused problems) and has many consequences, e.g. the imposing of Dirichlet bcs for a mixed method with a monolithic matrix would need to change somehow.

I would welcome concrete suggestions for changes, but they do need to be considered in the round (monolithic matrix, block matrixes, blocked forms assembled into monolithic matrix, assembling off-diagonal and diagonal block separately, ...) and not just for an isolated case.

A change I have considered previously is raising an exception if a boundary condition that is passed to the assembler is not used. We could also pass bcs as separate test (row) and trial (column) lists.

**Comment by wence-** [comment]
We use a process where you might say something like:

```
V = FunctionSpace(...)
W = V*V

bc1 = DirichletBC(W.sub(0), ...)
bc2 = DirichletBC(W.sub(1), ...)

assemble(a, bcs=[bc1, bc2])
```

Now `W.sub` produces an `IndexedFunctionSpace` that is just, effectively a tuple of `(V, index)`. So you shared the data of the underlying space (good), but can differentiate where `bc1` and `bc2` apply by their indices. This works for us because the mixed space is just a bag of spaces, and when inserting into a global matrix we arrange for different local-to-global maps to be applied when using `V.sub(0)` or `V.sub(1)`.

We've not thought about how to do application of BCs where the row and column BCs would be different.

**Comment by garth-wells** [comment]
Fix is in #626 

Also see block assembly tests.
